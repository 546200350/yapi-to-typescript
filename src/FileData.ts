export { FileData as default } from 'vtils'
