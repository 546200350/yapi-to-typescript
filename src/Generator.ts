import JSON5 from 'json5'
import { CategoryList, Config, Interface, InterfaceList, Method, PropDefinition, RequestBodyType, RequestFormItemType, Required, ResponseBodyType } from './types'
import { isEmpty } from 'vtils'
import { JSONSchema4 } from 'json-schema'
import { jsonSchemaStringToJsonSchema, jsonSchemaToType, jsonToJsonSchema, mockjsTemplateToJsonSchema, propDefinitionsToJsonSchema } from './utils'

export default class Generator {
  /** 配置 */
  private config: Config

  /** { 项目标识: 分类列表 } */
  private projectIdToCategoryList: Record<string, CategoryList | undefined> = Object.create(null)

  constructor(config: Config) {
    this.config = config
  }

  /** 生成请求数据类型 */
  static async generateRequestDataType(
    { interfaceInfo, typeName }: {
      interfaceInfo: Interface,
      typeName: string,
    },
  ): Promise<string> {
    let jsonSchema: JSONSchema4 = {}

    switch (interfaceInfo.method) {
      case Method.GET:
      case Method.HEAD:
      case Method.OPTIONS:
        jsonSchema = propDefinitionsToJsonSchema(
          interfaceInfo.req_query.map<PropDefinition>(item => ({
            name: item.name,
            required: item.required === Required.true,
            type: 'string',
            comment: item.desc,
          })),
        )
        break
      default:
        switch (interfaceInfo.req_body_type) {
          case RequestBodyType.form:
            jsonSchema = propDefinitionsToJsonSchema(
              interfaceInfo.req_body_form.map<PropDefinition>(item => ({
                name: item.name,
                required: item.required === Required.true,
                type: (item.type === RequestFormItemType.file ? 'file' : 'string') as any,
                comment: item.desc,
              })),
            )
            break
          case RequestBodyType.json:
            if (interfaceInfo.req_body_other) {
              jsonSchema = interfaceInfo.req_body_is_json_schema
                ? jsonSchemaStringToJsonSchema(interfaceInfo.req_body_other)
                : jsonToJsonSchema(JSON5.parse(interfaceInfo.req_body_other))
            }
            break
          default:
            break
        }
        break
    }

    return jsonSchemaToType(jsonSchema, typeName)
  }

  /** 生成响应数据类型 */
  static async generateResponseDataType(
    { interfaceInfo, typeName, dataKey }: {
      interfaceInfo: Interface,
      typeName: string,
      dataKey?: string,
    },
  ): Promise<string> {
    let jsonSchema: JSONSchema4 = {}

    switch (interfaceInfo.res_body_type) {
      case ResponseBodyType.json:
      case ResponseBodyType.jsonSchema:
        if (interfaceInfo.res_body) {
          jsonSchema = interfaceInfo.res_body_is_json_schema
            ? jsonSchemaStringToJsonSchema(interfaceInfo.res_body)
            : mockjsTemplateToJsonSchema(JSON5.parse(interfaceInfo.res_body))
        }
        break
      default:
        break
    }

    if (dataKey && jsonSchema && jsonSchema.properties && jsonSchema.properties[dataKey]) {
      jsonSchema = jsonSchema.properties[dataKey]
    }

    return jsonSchemaToType(jsonSchema, typeName)
  }

  /** 获取分类的接口列表 */
  async fetchCategoryInterfaceList(
    { serverUrl, token, categoryId }: {
      serverUrl: string,
      token: string,
      categoryId: number,
    },
  ): Promise<InterfaceList> {
    const projectId: string = `${serverUrl}|${token}`

    if (!(projectId in this.projectIdToCategoryList)) {
      this.projectIdToCategoryList[projectId] = []
    }

    const category = (this.projectIdToCategoryList[projectId] || []).find(
      cat => (
        !isEmpty(cat)
          && !isEmpty(cat.list)
          && cat.list[0].catid === categoryId
      ),
    )

    return category ? category.list : []
  }
}
