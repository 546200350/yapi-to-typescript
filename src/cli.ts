#!/usr/bin/env node

import { run } from './cliCore'

(async () => {
  await run()
})()
