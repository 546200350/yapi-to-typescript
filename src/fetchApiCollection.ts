import request from 'superagent'
import { ApiCollection, Config } from './types'

type LoginResult = {
  body: {
    errcode: number,
    errmsg: string,
  },
}
type ApiResult = { body: Buffer }

export default async function fetchApiCollection(config: Config): Promise<ApiCollection> {
  const [, baseUrl, projectId] = config.projectUrl.match(/(.+\/)project\/(\d+)\//)

  const loginUrl = `${baseUrl}api/user/${config.login.method === 'ldap' ? 'login_by_ldap' : 'login'}`
  const apiUrl = `${baseUrl}api/plugin/export?type=json&pid=${projectId}&status=all&isWiki=false`

  const agent = request.agent()

  if (config.extraCookies) {
    (agent as any).set('Cookie', config.extraCookies)
  }

  const { body: loginResult } = await agent
    .post(loginUrl)
    .send({
      email: config.login.email,
      password: config.login.password,
    }) as LoginResult
  if (!loginResult || loginResult.errcode !== 0) {
    throw new Error(
      `\x1b[31m登录 ${loginUrl} 失败，请检查邮箱、密码是否有误或服务是否可用。（服务器错误信息：${loginResult.errmsg}）\x1b[0m`
    )
  }

  const { body: api } = await agent.get(apiUrl).buffer() as ApiResult

  return JSON.parse(api.toString()) as any
}
