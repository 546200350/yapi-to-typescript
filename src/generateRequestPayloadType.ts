import { JSONSchema4 } from 'json-schema'
// @ts-ignore
import jsonSchemaGenerator from 'json-schema-generator'
import JSON5 from 'json5'
import { Api, Method, Required, RequestBodyType, RequestFormItemType, PropDefinition } from './types'
import propDefinitionsToJsonSchema from './propDefinitionsToJsonSchema'
import jsonSchemaToTypes from './jsonSchemaToTypes'

/**
 * 生成请求数据的 typescript 类型定义。
 *
 * @param api Api
 * @param interfaceName 接口名称
 * @returns typescript 定义
 */
export default async function generateRequestPayloadType(api: Api, interfaceName: string): Promise<string> {
  let jsonSchema: JSONSchema4 = {}
  switch (api.method) {
    case Method.GET:
    case Method.HEAD:
    case Method.OPTIONS:
      jsonSchema = propDefinitionsToJsonSchema(
        api.req_query.map<PropDefinition>(item => ({
          name: item.name,
          required: item.required === Required.True,
          type: 'string',
          comment: item.desc,
        })),
      )
      break
    default:
      switch (api.req_body_type) {
        case RequestBodyType.Form:
          jsonSchema = propDefinitionsToJsonSchema(
            api.req_body_form.map<PropDefinition>(item => ({
              name: item.name,
              required: item.required === Required.True,
              type: (item.type === RequestFormItemType.File ? 'file' : 'string') as any,
              comment: item.desc,
            })),
          )
          break
        case RequestBodyType.Json:
          if (api.req_body_other) {
            jsonSchema = api.req_body_is_json_schema
              ? JSON.parse(api.req_body_other)
              : jsonSchemaGenerator(JSON5.parse(api.req_body_other))
          }
          break
        default:
          break
      }
      break
  }
  return jsonSchemaToTypes(jsonSchema, interfaceName)
}
