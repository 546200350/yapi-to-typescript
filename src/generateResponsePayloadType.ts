import { JSONSchema4 } from 'json-schema'
import JSON5 from 'json5'
import { Api, ResponseBodyType } from './types'
import mockjsJsonToJsonSchema from './mockjsJsonToJsonSchema'
import jsonSchemaToTypes from './jsonSchemaToTypes'

/**
 * 生成返回数据的 typescript 类型定义。
 *
 * @param api Api
 * @param interfaceName 接口名称
 * @param [dataKey] 数据所在字段，不设置表示整体都是数据
 * @returns typescript 定义
 */
export default async function generateResponsePayloadType(api: Api, interfaceName: string, dataKey?: string): Promise<string> {
  let jsonSchema: JSONSchema4
  switch (api.res_body_type) {
    case ResponseBodyType.Json:
    case ResponseBodyType.JsonSchema:
      if (api.res_body_is_json_schema) {
        jsonSchema = api.res_body && JSON.parse(api.res_body)
      } else {
        jsonSchema = api.res_body && mockjsJsonToJsonSchema(JSON5.parse(api.res_body))
      }
      break
    default:
      break
  }
  if (dataKey && jsonSchema && jsonSchema.properties && jsonSchema.properties[dataKey]) {
    jsonSchema = jsonSchema.properties[dataKey]
  }
  return jsonSchemaToTypes(jsonSchema, interfaceName)
}
