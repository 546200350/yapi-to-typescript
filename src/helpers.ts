export class FileData<T = any> {
  /**
   * 原始文件数据。
   */
  private originalFileData: T

  /**
   * 文件数据辅助类，统一网页、小程序等平台的文件上传。
   *
   * @param originalFileData 原始文件数据
   */
  public constructor(originalFileData: T) {
    this.originalFileData = originalFileData
  }

  /**
   * 获取原始文件数据。
   *
   * @returns 原始文件数据
   */
  public getOriginalFileData(): T {
    return this.originalFileData
  }
}

/**
 * 解析请求数据，从请求数据中分离出普通数据和文件数据。
 *
 * @param [requestData] 要解析的请求数据
 * @returns 包含普通数据(data)和文件数据(fileData)的对象，data、fileData 为空对象时，表示没有此类数据
 */
export function parseRequestData(requestData?: any): { data: any, fileData: any } {
  const result = {
    data: {} as any,
    fileData: {} as any,
  }
  if (requestData != null) {
    if (typeof requestData === 'object' && !Array.isArray(requestData)) {
      Object.keys(requestData).forEach(key => {
        if (requestData[key] && requestData[key] instanceof FileData) {
          result.fileData[key] = (requestData[key] as FileData).getOriginalFileData()
        } else {
          result.data[key] = requestData[key]
        }
      })
    } else {
      result.data = requestData
    }
  }
  return result
}
