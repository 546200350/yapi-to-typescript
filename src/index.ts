import path from 'path'
import fs from 'fs-extra'
import * as changeCase from 'change-case'
import consola from 'consola'
import { Config, ApiList, InterfaceType, ExtendedApi } from './types'
import fetchApiCollection from './fetchApiCollection'
import generateRequestPayloadType from './generateRequestPayloadType'
import generateResponsePayloadType from './generateResponsePayloadType'

export default async (config: Config): Promise<void> => {
  consola.info('获取接口 JSON 文件中...')
  // 获取接口的 JSON 文件并转换为对象
  const apiCollection = await fetchApiCollection(config)
  consola.info('生成 TypeScript 类型文件中...')
  // 生成分类 ID 到分类 API 列表的对象
  const categoryIdToApiList = apiCollection.reduce<{ [id: number]: ApiList }>((res, api) => {
    if (api.list.length) {
      res[api.list[0].catid] = api.list
    }
    return res
  }, {})
  const tsContent = (
    await Promise.all(
      Object.keys(config.categories).map(async (categoryId: any) => {
        const { getRequestFunctionName, getInterfaceName } = config.categories[categoryId]
        return Promise.all(
          categoryIdToApiList[categoryId].map(async api => {
            const extendedApi: ExtendedApi = {
              ...api,
              parsedPath: path.parse(api.path),
              changeCase: changeCase,
            }
            const requestDataInterfaceName = changeCase.pascalCase(getInterfaceName(extendedApi, InterfaceType.Request))
            const responseDataInterfaceName = changeCase.pascalCase(getInterfaceName(extendedApi, InterfaceType.Response))
            const requestPayloadType = (await generateRequestPayloadType(api, requestDataInterfaceName)).trim()
            const responsePayloadType = (await generateResponsePayloadType(api, responseDataInterfaceName, config.dataKey)).trim()
            return [
              `/**\n * **请求类型**：${api.title}\n */\n${requestPayloadType}`,
              `/**\n * **响应类型**：${api.title}\n */\n${responsePayloadType}`,
              `/**\n * ${api.title}\n */\nexport function ${getRequestFunctionName(extendedApi)}(requestData${/(\{\}|any)$/s.test(requestPayloadType) ? '?' : ''}: ${requestDataInterfaceName}): Promise<${responseDataInterfaceName}> {\n${
                [
                  `  const { data, fileData } = parseRequestData(requestData)`,
                  `  return request({`,
                  `    path: '${api.path}',`,
                  `    method: '${api.method}',`,
                  `    requestBodyType: '${api.req_body_type}',`,
                  `    responseBodyType: '${api.res_body_type}',`,
                  `    data: data,`,
                  `    fileData: fileData`,
                  `  } as any)`,
                ].join('\n')
              }\n}`,
            ].join('\n\n')
          })
        )
      })
    )
  )
    .reduce((res, arr) => {
      res.push(...arr)
      return res
    }, [])
    .join('\n\n')
  const targetFile = path.resolve(process.cwd(), config.targetFile)
  const requestFile = path.join(path.parse(targetFile).dir, 'request.ts')
  if (!fs.existsSync(requestFile)) {
    fs.outputFileSync(requestFile, `${`
      import { RequestPayload } from 'yapi-to-typescript/lib/types'

      export default function request({
        path,
        method,
        requestBodyType,
        responseBodyType,
        data,
        fileData
      }: RequestPayload): Promise<any> {
        return new Promise((resolve, reject) => {
          // 是否含有文件数据
          const hasFileData = fileData !== null

          // 在这里实现请求逻辑
        })
      }
    `.trim().replace(/ {6}/g, '')}\n`)
  }
  fs.outputFileSync(targetFile, [
    `/* tslint:disable */\n/* eslint-disable */`,
    `import request from './request'`,
    `// @ts-ignore\nimport { FileData, parseRequestData } from 'yapi-to-typescript/lib/utils'`,
    tsContent,
  ].join('\n\n'))
  consola.success(`操作完成.`)
}
