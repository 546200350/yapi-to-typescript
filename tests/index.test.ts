import path from 'path'
import fs from 'fs-extra'
import ytt from '../src'
import { InterfaceType } from '../src/types'

const tmpDir = path.join(__dirname, '.tmp1')
const targetFile = path.join(tmpDir, 'api/index.ts')
const requestFile = path.join(tmpDir, 'api/request.ts')

beforeAll(() => {
  fs.ensureDirSync(tmpDir)
})

afterAll(() => {
  fs.removeSync(tmpDir)
})

test('成功', async () => {
  await ytt({
    projectUrl: 'http://foo.bar/project/20/interface/api',
    email: 'hello@x.xx',
    password: 'correctPassword',
    targetFile: targetFile,
    categories: {
      58: {
        getInterfaceName({ changeCase, parsedPath }, interfaceType) {
          return changeCase.camelCase(parsedPath.name) + (interfaceType === InterfaceType.Request ? 'Request' : 'Response')
        },
        getRequestFunctionName({ changeCase, parsedPath }) {
          return changeCase.camelCase(parsedPath.name)
        },
      },
    },
  })
  expect(fs.readFileSync(targetFile).toString()).toMatchSnapshot()
  expect(fs.readFileSync(requestFile).toString()).toMatchSnapshot()
})
