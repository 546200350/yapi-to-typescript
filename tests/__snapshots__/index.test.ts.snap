// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`成功 1`] = `
"/* tslint:disable */
/* eslint-disable */

import request from './request'

// @ts-ignore
import { FileData, parseRequestData } from 'yapi-to-typescript/lib/utils'

/**
 * **请求类型**：get
 */
export interface GetRequest {}

/**
 * **响应类型**：get
 */
export interface GetResponse {
  data: string;
}

/**
 * get
 */
export function get(requestData?: GetRequest): Promise<GetResponse> {
  const { data, fileData } = parseRequestData(requestData)
  return request({
    path: '/get',
    method: 'GET',
    requestBodyType: 'undefined',
    responseBodyType: 'json',
    data: data,
    fileData: fileData
  } as any)
}

/**
 * **请求类型**：post
 */
export interface PostRequest {
  page: number;
}

/**
 * **响应类型**：post
 */
export interface PostResponse {
  list: string[];
}

/**
 * post
 */
export function post(requestData: PostRequest): Promise<PostResponse> {
  const { data, fileData } = parseRequestData(requestData)
  return request({
    path: '/post',
    method: 'POST',
    requestBodyType: 'json',
    responseBodyType: 'json',
    data: data,
    fileData: fileData
  } as any)
}"
`;

exports[`成功 2`] = `
"import { RequestPayload } from 'yapi-to-typescript/lib/types'

export default function request({
  path,
  method,
  requestBodyType,
  responseBodyType,
  data,
  fileData
}: RequestPayload): Promise<any> {
  return new Promise((resolve, reject) => {
    // 是否含有文件数据
    const hasFileData = fileData !== null

    // 在这里实现请求逻辑
  })
}
"
`;
